design_docs/2023-02_hydro_cli.md

```diff
@@ -0,0 +1,105 @@
+```rust
+// src/echo_server.hf
+use data_types::EchoMsg; // defined in data_types crate in this workspace
+use chrono::prelude::*;
+use std::net::SocketAddr;
+
+module main(
+    input inbound: (EchoMsg, SocketAddr),
+    output outbound: (EchoMsg, SocketAddr)
+) {
+    // Define a shared inbound channel
+    inbound_chan = inbound -> tee();
+
+    // Print all messages for debugging purposes
+    inbound_chan[0]
+        -> for_each(|(msg, addr): (EchoMsg, SocketAddr)| println!("{}: Got {:?} from {:?}", Utc::now(), msg, addr));
+
+    // Echo back the Echo messages with updated timestamp
+    inbound_chan[1]
+        -> map(|(EchoMsg {payload, ..}, addr)| (EchoMsg { payload, ts: Utc::now() }, addr) ) -> outbound;
+}
+```
+
+Assume we also have a `src/echo_client.hf` file that periodically sends `EchoMsg`s to an `outbound` channel and logs echos from the `inbound` channel. We can deploy this program to a cloud machine using the following `echo.hydro.py` config file:
+```python
+from hydro import Deployment
+from hydro.gcp import GCPMachine # keys are automatically loaded from somewhere
+
+async def main():
+    deployment = Deployment()
+
+    # Specify the GCP instances we want to deploy to
+    server_machine = deployment.GCPMachine(
+        project="hydro-1234",
+        zone="us-west1-a",
+        type="e2-micro",
+    )
+
+    client_machine = deployment.GCPMachine(
+        project="hydro-1234",
+        zone="us-west1-a",
+        type="e2-micro",
+    )
+
+    # Load Hydroflow programs
+    server_hf = deployment.Hydroflow(
+        src="src/echo_server.hf",
+        on=server_machine,
+    )
+
+    client_hf = deployment.Hydroflow(
+        src="src/echo_client.hf",
+        on=client_machine,
+    )
+
+    # Wire the programs
+    client_hf.ports.outbound.connect(server_hf.ports.inbound)
+    server_hf.ports.outbound.connect(client_hf.ports.inbound)
+
+    # Launch the two programs
+    await deployment.deploy()
+
+    # Once deployment finishes, start capturing logs from the client
+    # (logs are buffered even before we grab `stdout`)
+    client_logs = client_hf.stdout()
+
+    i = 0
+    async for log in client_logs:
+        print(log)
+        i += 1
+        if i == 10:
+            break
+
+    # Stop the programs (optional)
+    await client_hf.stop()
+    await server_hf.stop()
+
+    # Kill the machines (optional)
+    await client_machine.terminate()
+    await server_machine.terminate()
+```
+
+Then, we can launch this deployment with a simple command:
+```bash
+$ hydro run echo.hydro.py
+```
+
+Which will automatically compile the Hydroflow programs for the target machine, launch the cloud machines, allocate internal ports for each connected input/output, and start the services.
+
+## Standalone Hydroflow
+This will eventually be a design document of its own, but for now we briefly lay out a prototype of how to augment Hydroflow programs to be meaningful without a surrounding Rust program.
+
+TODO
```

Comment:
How does this relate to https://github.com/hydro-project/hydroplane

Reply:
I see Hydroplane as a backend for the provisioning side of the CLI. Right now I'm planning on using Terraform at first so we can also support raw machines, but Hydroplane can be integrated similarly.

Reply:
Please don't start leaking Terraform into our stuff. If we're doing deployment we should use Hydroplane, and if we don't like it we should fix it.

Reply:
I think the tricky part is we do want support for deploying raw machines, which means either integrating with each provider ourselves or using some abstraction like Terraform.

And it's nice to expose Rust to Python rather than the other way around so my general plan is to use Hydroplane's existing capabilities here but in the long term merge things in so that we have new capabilities implemented in Rust.